doc/man/doveadm-instance.1.in
.\" Copyright (c) 2012-2018 Dovecot authors, see the included COPYING file
.TH DOVEADM\-INSTANCE 1 "2012-02-16" "Dovecot v2.3" "Dovecot"
.SH NAME
doveadm\-instance \- Manage the list of running Dovecot instances
.\"------------------------------------------------------------------------
.SH SYNOPSIS
.BR doveadm " [" \-Dv "] [" \-f
.IR formatter ]
.B instance list
.br
.BR doveadm " [" \-Dv "] " "instance remove"
.IR name " | " base_dir
.\"------------------------------------------------------------------------
.SH DESCRIPTION
The
.B doveadm instance
commands are used to manage the list of Dovecot instances running on the
server.
In most installations there is only one Dovecot instance, but in some cases
is may be useful to have more (e.g. running director proxy and backend in
the same server).
.PP
Instances are added to the list automatically when Dovecot is started.
Each instance is uniquely identified by its
.I base_dir
setting.
Instances can be named by setting
.I instance_name
in each instance\(aqs
.IR dovecot.conf .
When an instance is named, it can be accessed easily by giving
.BI \-i\  instance_name
command line parameter for Dovecot binaries (e.g. doveadm).
.\"------------------------------------------------------------------------
@INCLUDE:global-options-formatter@
.\"------------------------------------------------------------------------
.SH ARGUMENTS
.TP
.I name
The value of an instance\(aqs
.I instance_name
setting.
.\"-------------------------------------
.TP
.I base_dir
The base directory of a Dovecot instance.
.\"------------------------------------------------------------------------
.SH COMMANDS
.SS instance list
.B doveadm instance list
.PP
This command lists the seen Dovecot instances.
.\"-------------------------------------
.SS instance remove
.B doveadm instance remove
.IR name " | " base_dir
.PP
This command removes the specified instance.
.\"------------------------------------------------------------------------
@INCLUDE:reporting-bugs@
.\"------------------------------------------------------------------------
.SH SEE ALSO
.BR doveadm (1)

NEWS
v2.3.1 2018-02-29  Aki Tuomi <aki.tuomi@dovecot.fi>

	* Submission server support improvements and bug fixes
	  - Lots of bug fixes to submission server
	* API CHANGE: array_idx_modifiable will no longer allocate space
	  - Particularly affects how you should check MODULE_CONTEXT result, or
	    use REQUIRE_MODULE_CONTEXT.

	+ mail_attachment_detection_options setting controls when
	  $HasAttachment and $HasNoAttachment keywords are set for mails.
	+ imap: Support fetching body snippets using FETCH (SNIPPET) or
	  (SNIPPET (LAZY=FUZZY))
	+ fs-compress: Automatically detect whether input is compressed or not.
	  Prefix the compression algorithm with "maybe-" to enable the
	  detection, for example: "compress:maybe-gz:6:..."
	+ Added settings to change dovecot.index* files' optimization behavior.
	  See https://wiki2.dovecot.org/IndexFiles#Settings
	+ Auth cache can now utilize auth workers to do password hash
	  verification by setting auth_cache_verify_password_with_worker=yes.
	+ Added charset_alias plugin. See
	  https://wiki2.dovecot.org/Plugins/CharsetAlias
	+ imap_logout_format and pop3_logout_format settings now support all of
	  the generic variables (e.g. %{rip}, %{session}, etc.)
	+ Added auth_policy_check_before_auth, auth_policy_check_after_auth
	  and auth_policy_report_after_auth settings.
	+ master: Support HAProxy PP2_TYPE_SSL command and set "secured"
	  variable appropriately
	- Invalid UCS4 escape in HTML can cause crashes
	- imap: IMAP COMPRESS -enabled clietn crashes on disconnect
	- lmtp: Fix crash when user is over quota
	- lib-lda: Parsing Return-Path header address fails when it contains
	  CFWS
	- auth: SASL with Exim fails for AUTH commands without an initial
	  response
	- imap: SPECIAL-USE capability isn't automatically added
	- auth: LDAP subqueries do not support standard auth variables in
	  var-expand
	- auth: SHA256-CRYPT and SHA512-CRYPT schemes do not work
	- lib-index: mail_always/never_cache_fields are not used for existing
	  cache files
	- imap: Fetching headers leaks memory if search doesn't find any mails
	- lmtp: ORCPT support in RCPT TO
	- imap-login: Process sometimes ends up in infinite loop
	- sdbox: Rolled back save/copy transaction doesn't delete temp files
	- mail: lock_method=dotlock causes crashes

v2.3.0.1 2018-02-28  Timo Sirainen <tss@iki.fi>

	* CVE-2017-15130: TLS SNI config lookups may lead to excessive
	  memory usage, causing imap-login/pop3-login VSZ limit to be reached
	  and the process restarted. This happens only if Dovecot config has
	  local_name { } or local { } configuration blocks and attacker uses
	  randomly generated SNI servernames.
	* CVE-2017-14461: Parsing invalid email addresses may cause a crash or
	  leak memory contents to attacker. For example, these memory contents
	  might contain parts of an email from another user if the same imap
	  process is reused for multiple users. First discovered by Aleksandar
	  Nikolic of Cisco Talos. Independently also discovered by "flxflndy"
	  via HackerOne.
	* CVE-2017-15132: Aborted SASL authentication leaks memory in login
	  process.
	* Linux: Core dumping is no longer enabled by default via
	  PR_SET_DUMPABLE, because this may allow attackers to bypass
	  chroot/group restrictions. Found by cPanel Security Team. Nowadays
	  core dumps can be safely enabled by using "sysctl -w
	  fs.suid_dumpable=2". If the old behaviour is wanted, it can still be
	  enabled by setting:
	  import_environment=$import_environment PR_SET_DUMPABLE=1
	- imap-login with SSL/TLS connections may end up in infinite loop

v2.3.0 2017-12-22  Timo Sirainen <tss@iki.fi>

	* Various setting changes, see https://wiki2.dovecot.org/Upgrading/2.3
	* Logging rewrite started: Logging is now based on hierarchical events.
	  This makes it possible to do various things, like: 1) giving
	  consistent log prefixes, 2) enabling debug logging with finer
	  granularity, 3) provide logs in more machine readable formats
	  (e.g. json). Everything isn't finished yet, especially a lot of the
	  old logging code still needs to be translated to the new way.
	* Statistics rewrite started: Stats are now based on (log) events.
	  It's possible to gather statistics about any event that is logged.
	  See http://wiki2.dovecot.org/Statistics for details
	* ssl_dh setting replaces the old generated ssl-parameters.dat
	* IMAP: When BINARY FETCH finds a broken mails, send [PARSE] error
	  instead of [UNKNOWNCTE]
	* Linux: core dumping via PR_SET_DUMPABLE is no longer enabled by
	  default due to potential security reasons (found by cPanel Security
	  Team).

	+ Added support for SMTP submission proxy server, which includes
	  support for BURL and CHUNKING extension.
	+ LMTP rewrite. Supports now CHUNKING extension and mixing of
	  local/proxy recipients.
	+ auth: Support libsodium to add support for ARGON2I and ARGON2ID
	  password schemes.
	+ auth: Support BLF-CRYPT password scheme in all platforms
	+ auth: Added LUA scripting support for passdb/userdb.
	  See https://wiki2.dovecot.org/AuthDatabase/Lua
	- Input streams are more reliable now when there are errors or when
	  the maximum buffer size is reached. Previously in some situations
	  this could have caused Dovecot to try to read already freed memory.
	- Output streams weren't previously handling failures when writing a
	  trailer at the end of the stream. This mainly affected encrypt and
	  zlib compress ostreams, which could have silently written truncated
	  files if the last write happened to fail (which shouldn't normally
	  have ever happened).
	- virtual plugin: Fixed panic when fetching mails from virtual
	  mailboxes with IMAP BINARY extension.
	- doveadm-server: Fix potential hangs with SSL connections
	- doveadm proxy: Reading commands' output from v2.2.33+ servers could
	  have caused the output to be corrupted or caused a crash.
	- Many other smaller fixes

v2.2.33.2 2017-10-20  Timo Sirainen <tss@iki.fi>


src/doveadm/doveadm-instance.c
/* Copyright (c) 2013-2018 Dovecot authors, see the included COPYING file */

#include "lib.h"
#include "master-instance.h"
#include "master-service-settings.h"
#include "doveadm.h"
#include "doveadm-print.h"

#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <signal.h>

extern struct doveadm_cmd doveadm_cmd_instance[];

static void instance_cmd_help(doveadm_command_t *cmd) ATTR_NORETURN;

static bool pid_file_read(const char *path)
{
	char buf[32];
	int fd;
	ssize_t ret;
	pid_t pid;
	bool found = FALSE;

	fd = open(path, O_RDONLY);
	if (fd == -1) {
		if (errno != ENOENT)
			i_error("open(%s) failed: %m", path);
		return FALSE;
	}

	ret = read(fd, buf, sizeof(buf));
	if (ret < 0)
		i_error("read(%s) failed: %m", path);
	else if (ret > 0 && buf[ret-1] == '\n') {
		buf[ret-1] = '\0';
		if (str_to_pid(buf, &pid) == 0) {
			found = !(pid == getpid() ||
				  (kill(pid, 0) < 0 && errno == ESRCH));
		}
	}
	i_close_fd(&fd);
	return found;
}

static void cmd_instance_list(int argc, char *argv[])
{
	struct master_instance_list *list;
	struct master_instance_list_iter *iter;
	const struct master_instance *inst;
	const char *instance_path, *pidfile_path;
	bool show_config = FALSE;
	int c;

	while ((c = getopt(argc, argv, "c")) > 0) {
		switch (c) {
		case 'c':
			show_config = TRUE;
			break;
		default:
			help(&doveadm_cmd_instance[0]);
		}
	}
	argv += optind;

	if (!show_config) {
		doveadm_print_init(DOVEADM_PRINT_TYPE_TABLE);
		doveadm_print_header("path", "path", DOVEADM_PRINT_HEADER_FLAG_EXPAND);
		doveadm_print_header_simple("name");
		doveadm_print_header_simple("last used");
		doveadm_print_header_simple("running");
	}

	instance_path = t_strconcat(service_set->state_dir,
				    "/"MASTER_INSTANCE_FNAME, NULL);
	list = master_instance_list_init(instance_path);
	iter = master_instance_list_iterate_init(list);
	while ((inst = master_instance_iterate_list_next(iter)) != NULL) {
		if (argv[0] != NULL && strcmp(argv[0], inst->name) != 0)
			continue;

		if (show_config) {
			printf("%s\n", inst->config_path == NULL ? "" :
			       inst->config_path);
			continue;
		}
		doveadm_print(inst->base_dir);
		doveadm_print(inst->name);
		doveadm_print(unixdate2str(inst->last_used));
		pidfile_path = t_strconcat(inst->base_dir, "/master.pid", NULL);
		if (pid_file_read(pidfile_path))
			doveadm_print("yes");
		else
			doveadm_print("no");
	}
	master_instance_iterate_list_deinit(&iter);
	master_instance_list_deinit(&list);
}

static void cmd_instance_remove(int argc, char *argv[])
{
	struct master_instance_list *list;
	const struct master_instance *inst;
	const char *base_dir, *instance_path;
	int ret;

	if (argc != 2)
		instance_cmd_help(cmd_instance_remove);

	instance_path = t_strconcat(service_set->state_dir,
				    "/"MASTER_INSTANCE_FNAME, NULL);
	list = master_instance_list_init(instance_path);
	inst = master_instance_list_find_by_name(list, argv[1]);
	base_dir = inst != NULL ? inst->base_dir : argv[1];
	if ((ret = master_instance_list_remove(list, base_dir)) < 0) {
		i_error("Failed to remove instance");
		doveadm_exit_code = EX_TEMPFAIL;
	} else if (ret == 0) {
		i_error("Instance already didn't exist");
		doveadm_exit_code = DOVEADM_EX_NOTFOUND;
	}
	master_instance_list_deinit(&list);
}

struct doveadm_cmd doveadm_cmd_instance[] = {
	{ cmd_instance_list, "instance list", "[-c] [<name>]" },
	{ cmd_instance_remove, "instance remove", "<name> | <base dir>" }
};

static void instance_cmd_help(doveadm_command_t *cmd)
{
	unsigned int i;

	for (i = 0; i < N_ELEMENTS(doveadm_cmd_instance); i++) {
		if (doveadm_cmd_instance[i].cmd == cmd)
			help(&doveadm_cmd_instance[i]);
	}
	i_unreached();
}

void doveadm_register_instance_commands(void)
{
	unsigned int i;

	for (i = 0; i < N_ELEMENTS(doveadm_cmd_instance); i++)
		doveadm_register_cmd(&doveadm_cmd_instance[i]);
}
